### PR TITLE
[Channel] Cached per request channel context

### PR DESCRIPTION
**Pull request by pamil**
Based on #4056.

### CODE REVIEW

File: src/Sylius/Bundle/ChannelBundle/Resources/config/services.xml [hunk 1]
```diff
@@ -11,17 +11,38 @@
...
     <parameters>
-        <parameter key="sylius.context.channel.class">Sylius\Component\Channel\Context\ChannelContext</parameter>
+        <parameter key="sylius.context.channel.composite.class">Sylius\Component\Channel\Context\CompositeChannelContext</parameter>
+
+        <parameter key="sylius.context.channel.request_based.class">Sylius\Component\Channel\Context\RequestBased\ChannelContext</parameter>
+        <parameter key="sylius.context.channel.single_channel.class">Sylius\Component\Channel\Context\SingleChannelContext</parameter>
+
+        <parameter key="sylius.context.channel.request_based.resolver.composite.class">Sylius\Component\Channel\Context\RequestBased\CompositeRequestResolver</parameter>
+        <parameter key="sylius.context.channel.request_based.resolver.hostname_based.class">Sylius\Component\Channel\Context\RequestBased\HostnameBasedRequestResolver</parameter>
+
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ChannelBundle/Resources/config/services.xml]: Redundant blank line

File: src/Sylius/Component/Channel/Context/CachedPerRequestChannelContext.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        $this->requestToChannelMap = new \SplObjectStorage();
+        $this->nullObject = new \stdClass();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getChannel()
+    {
+        $objectIdentifier = $this->requestStack->getMasterRequest() ?: $this->nullObject;
```
Comment (lchrusciel) [inline src/Sylius/Component/Channel/Context/CachedPerRequestChannelContext.php]: Why do we need reference to this null object? Won't it be better to just create new `\stdClass()` instance here?

Comment (pamil) [inline src/Sylius/Component/Channel/Context/CachedPerRequestChannelContext.php]: The objects are compared by references.

**Comment by michalmarcinkowski** [comment]
:+1:

**Comment by pjedrzejewski** [comment]
Thank you Kamil! :+1:

**Pull request merged**